WooCommerce/Classes/ViewRelated/Dashboard/Onboarding/StoreOnboardingLaunchStoreViewModel.swift

```diff
@@ -3,41 +3,100 @@ import Yosemite
 
 /// View model for `StoreOnboardingLaunchStoreView`.
 final class StoreOnboardingLaunchStoreViewModel: ObservableObject {
+    /// UI state of the lauch store view.
+    enum State {
+        // Checking current site's plan to check for free trial
+        case checkingSitePlan
+        // Using free trial. Need to purchase plan before publishing site.
+        case needsPlanUpgrade
+        // Ready to publish site.
+        case readyToPublish
+        // Processing launch store request
+        case launchingStore
+    }
+
     let siteURL: URL
 
-    @Published private(set) var isLaunchingStore: Bool = false
+    @Published private(set) var state: State = .checkingSitePlan
     @Published var error: SiteLaunchError?
 
     private let siteID: Int64
     private let stores: StoresManager
     private let onLaunch: () -> Void
 
+    /// Closure invoked when the merchants taps on the `Upgrade` button.
+    ///
+    private let onUpgradeTapped: () -> Void
+
     init(siteURL: URL,
          siteID: Int64,
          stores: StoresManager = ServiceLocator.stores,
-         onLaunch: @escaping () -> Void) {
+         onLaunch: @escaping () -> Void,
+         onUpgradeTapped: @escaping () -> Void) {
         self.siteURL = siteURL
         self.siteID = siteID
         self.stores = stores
         self.onLaunch = onLaunch
+        self.onUpgradeTapped = onUpgradeTapped
+    }
+
+    @MainActor
+    func checkEligibilityToPublishStore() async {
+        update(state: .checkingSitePlan)
+        if await isFreeTrialPlan() {
+            update(state: .needsPlanUpgrade)
+        } else {
+            update(state: .readyToPublish)
+        }
     }
 
     @MainActor
     func launchStore() async {
         error = nil
-        isLaunchingStore = true
+        update(state: .launchingStore)
         let result = await launchStoreRemotely()
         switch result {
         case .success:
             onLaunch()
         case .failure(let error):
             self.error = error
+            update(state: .readyToPublish)
         }
-        isLaunchingStore = false
+    }
+
+    @MainActor
+    func didTapUpgrade() {
+        onUpgradeTapped()
     }
 }
 
 private extension StoreOnboardingLaunchStoreViewModel {
+    @MainActor
+    func update(state: State) {
+        self.state = state
+    }
+
+    @MainActor
+    func isFreeTrialPlan() async -> Bool {
```

Comment:
Super nit: `isFreeTrialPlan` sounds more like a variable - please consider making this a computed variable or renaming it to `checkFreeTrialPlan`.

Reply:
Done in 65d371d3bb726acb6e8beaa3ee4862ad43a28f0a